File: android/app/src/main/java/com/aitsuki/liveness/sample/LivenessUtils.kt
package com.aitsuki.liveness.sample

import android.graphics.PointF
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceContour
import com.google.mlkit.vision.face.FaceLandmark
import kotlin.math.abs
import kotlin.math.acos
import kotlin.math.sqrt


object LivenessUtils {

    private const val YAW_THRESHOLD = 12f
    private const val PITCH_THRESHOLD = 8f
    private const val ROLL_THRESHOLD = 8f
    private const val SIDE_FACE_YAW_THRESHOLD = 20f

    fun isFrontFace(face: Face): Boolean {
        val yaw = face.headEulerAngleY // 左右摇头角度
        val pitch = face.headEulerAngleX // 上下点头角度
        val roll = face.headEulerAngleZ // 旋转角度
        return yaw < YAW_THRESHOLD && yaw > -YAW_THRESHOLD
                && pitch < PITCH_THRESHOLD && pitch > -PITCH_THRESHOLD
                && roll < ROLL_THRESHOLD && roll > -ROLL_THRESHOLD
    }

    fun isSideFace(face: Face): Boolean {
        val yaw = face.headEulerAngleY // 左右摇头角度
        val pitch = face.headEulerAngleX // 上下点头角度
        val roll = face.headEulerAngleZ // 旋转角度
        return (yaw > SIDE_FACE_YAW_THRESHOLD || yaw < -SIDE_FACE_YAW_THRESHOLD)
                && pitch < PITCH_THRESHOLD && pitch > -PITCH_THRESHOLD
                && roll < ROLL_THRESHOLD && roll > -ROLL_THRESHOLD
    }

    fun isMouthOpened(face: Face): Boolean {
        val left = face.getLandmark(FaceLandmark.MOUTH_LEFT)?.position ?: return false
        val right = face.getLandmark(FaceLandmark.MOUTH_RIGHT)?.position ?: return false
        val bottom = face.getLandmark(FaceLandmark.MOUTH_BOTTOM)?.position ?: return false

        // Square of lengths be a2, b2, c2
        val a2 = lengthSquare(right, bottom)
        val b2 = lengthSquare(left, bottom)
        val c2 = lengthSquare(left, right)

        // length of sides be a, b, c
        val a = sqrt(a2)
        val b = sqrt(b2)

        // From Cosine law
        val gamma = acos((a2 + b2 - c2) / (2 * a * b))

        // Converting to degrees
        val gammaDeg = gamma * 180 / Math.PI
        return gammaDeg < 115f
    }

    private fun lengthSquare(a: PointF, b: PointF): Float {
        val x = a.x - b.x
        val y = a.y - b.y
        return x * x + y * y
    }

    /**
     * 远近检测
     * -1: tooClose
     * 0: perfect
     * 1: tooFar
     */
    fun isFaceTooFarOrClose(face: Face, imageWidth: Int, imageHeight: Int): Int {
        val boundingBox = face.boundingBox
        val contours = face.getContour(FaceContour.FACE)
        val top = contours?.points?.getOrNull(0)?.y?.toInt() ?: boundingBox.top
        val bottom = contours?.points?.getOrNull(18)?.y?.toInt() ?: boundingBox.bottom
        val left = contours?.points?.getOrNull(27)?.x?.toInt() ?: boundingBox.left
        val right = contours?.points?.getOrNull(9)?.x?.toInt() ?: boundingBox.right
        val height = bottom - top
        val width = right - left
        val widthPercent = width.toFloat() / imageWidth
        val heightPercent = height.toFloat() / imageHeight
        if (widthPercent > 0.8f || heightPercent > 0.8f) {
            return -1
        } else if (widthPercent < 0.3f || heightPercent < 0.3f) {
            return 1
        }
        return 0
    }

    /**
     * 居中检测
     */
    fun isFaceInCenter(face: Face, imageWidth: Int, imageHeight: Int): Boolean {
        val boundingBox = face.boundingBox
        val contours = face.getContour(FaceContour.FACE)
        val top = contours?.points?.getOrNull(0)?.y?.toInt() ?: boundingBox.top
        val bottom = contours?.points?.getOrNull(18)?.y?.toInt() ?: boundingBox.bottom
        val left = contours?.points?.getOrNull(27)?.x?.toInt() ?: boundingBox.left
        val right = contours?.points?.getOrNull(9)?.x?.toInt() ?: boundingBox.right
        val topMargin = top.coerceAtLeast(1)
        val bottomMargin = (imageHeight - bottom).coerceAtLeast(1)
        val leftMargin = left.coerceAtLeast(1)
        val rightMargin = (imageWidth - right).coerceAtLeast(1)
        val dh = abs(rightMargin - leftMargin)
        val dv = abs(bottomMargin - topMargin)
        return !(dh > imageWidth * 0.2f || dv > imageHeight * 0.2f)
    }
}

File: flutter/lib/camera/liveness_utils.dart
import 'dart:math';

import 'package:google_mlkit_face_detection/google_mlkit_face_detection.dart';

const double _yawThreshold = 12.0;
const double _pitchThreshold = 8.0;
const double _rollThreshold = 8.0;
const double _sideFaceYawThreshold = 20.0;

class LivenessUtils {
  LivenessUtils._();

  static bool isFrontFace(Face face) {
    final yaw = face.headEulerAngleY!; // 左右摇头角度
    final pitch = face.headEulerAngleX!; // 上下点头角度
    final roll = face.headEulerAngleZ!; // 旋转角度
    return yaw.abs() < _yawThreshold &&
        pitch.abs() < _pitchThreshold &&
        roll.abs() < _rollThreshold;
  }

  static bool isSideFace(Face face) {
    final yaw = face.headEulerAngleY!; // 左右摇头角度
    final pitch = face.headEulerAngleX!; // 上下点头角度
    final roll = face.headEulerAngleZ!; // 旋转角度
    return (yaw > _sideFaceYawThreshold || yaw < -_sideFaceYawThreshold) &&
        pitch.abs() < _pitchThreshold &&
        roll.abs() < _rollThreshold;
  }

  static bool isMouthOpened(Face face) {
    final left = face.landmarks[FaceLandmarkType.leftMouth]?.position;
    final right = face.landmarks[FaceLandmarkType.rightMouth]?.position;
    final bottom = face.landmarks[FaceLandmarkType.bottomMouth]?.position;

    if (left == null || right == null || bottom == null) {
      return false;
    }
    int a2 = right.squaredDistanceTo(bottom);
    int b2 = left.squaredDistanceTo(bottom);
    int c2 = left.squaredDistanceTo(right);
    double a = sqrt(a2);
    double b = sqrt(b2);
    double gamma = acos((a2 + b2 - c2) / (2 * a * b));
    double gammaDeg = gamma * 180 / pi;
    return gammaDeg < 115;
  }

  /// 远近检测
  /// -1: tooClose
  /// 0: perfect
  /// 1: tooFar
  static int isFaceTooFarOrClose(Face face, int imageWidth, int imageHeight) {
    final boundingBox = face.boundingBox;
    final contours = face.contours[FaceContourType.face];
    final top = contours?.points[0].y ?? boundingBox.top.toInt();
    final bottom = contours?.points[18].y ?? boundingBox.bottom.toInt();
    final left = contours?.points[27].x ?? boundingBox.left.toInt();
    final right = contours?.points[9].x ?? boundingBox.right.toInt();
    final height = bottom - top;
    final width = right - left;
    final widthPercent = width / imageWidth;
    final heightPercent = height / imageHeight;

    if (widthPercent > 0.9 || heightPercent > 0.9) {
      return -1; // too close
    } else if (widthPercent < 0.3 || heightPercent < 0.3) {
      return 1; // too far
    }
    return 0; // perfect
  }

  /// 居中检测
  static bool isFaceInCenter(Face face, int imageWidth, int imageHeight) {
    final boundingBox = face.boundingBox;
    final contours = face.contours[FaceContourType.face];
    final top = contours?.points[0].y ?? boundingBox.top.toInt();
    final bottom = contours?.points[18].y ?? boundingBox.bottom.toInt();
    final left = contours?.points[27].x ?? boundingBox.left.toInt();
    final right = contours?.points[9].x ?? boundingBox.right.toInt();
    final topMargin = max(top, 1);
    final bottomMargin = max(imageHeight - bottom, 1);
    final leftMargin = max(left, 1);
    final rightMargin = max(imageWidth - right, 1);
    final dh = (rightMargin - leftMargin).abs();
    final dv = (bottomMargin - topMargin).abs();
    return !(dh > imageWidth * 0.2 || dv > imageHeight * 0.2);
  }
}
